### ISSUE TITLE
ParseKconfig: unambiguous return value "true"

### ISSUE THREAD

**Issue by mythi** [kubernetes-sigs/node-feature-discovery#361]
ParseKconfig() is unambiguous if `CONFIG_FOOBAR="true"`. This results in the same return value as `CONFIG_FOOBAR=y` or `CONFIG_FOOBAR=m`

**Comment by marquiz** [comment]
I don't know if this is a real issue, though, as the config flag is either "tristate" or "string" type. I.e. CONFIG_FOO is either m/y or a string value. It cannot be `m`, `y` or `"true"`

**Comment by mythi** [comment]
@marquiz IIRC the problem was that you cannot distinguish between `m` and `y`.

**Comment by vaibhav2107** [comment]
Hi, I am new to this project and would love to contribute.
Can I work on this issue.

**Comment by marquiz** [comment]
Hi @vaibhav2107! Sure, you're very welcome to work on this.

I have been thinking how to fix this without breaking existing usage 🤔 I think one way to go would be to specify a separate type for tristate (i.e. *m/y* type) flags. Then change `ParseKconfig` to return `map[string]interface{}` and use a type switch in the caller's end to "do the right thing"

**Comment by vaibhav2107** [comment]
Thanks @marquiz for the guidance.

**Comment by marquiz** [comment]
This is actually somewhat nasty issue to tackle if we don't want to break backwards compatibility 🧐

In any case, I think that the underlying "raw" feature should be changed to just contain the actual value as a string. This would affect the new `kernel.config` rules where you'd need to do

```
        - feature: kernel.config
          matchExpressions:
            FOO: {op: In, value: ["y", "m"]}
```

instead of

```
        - feature: kernel.config
          matchExpressions:
            FOO: {op: IsTrue}
```

Then, the backwards compatibility, I see at least two options

### 1. `"y"` === `y`

This would be simple to implement, e.g. in the kernel source do something like this:

```
if rawVal == "y" || rawVal == "m" {
   labelVal = true
}
```

Correspondingly, in the custom source we would be need to add some magic to the legacy `kconfig` rule to handle the change.

This has the caveat the if some flag in kernel config would be `CONFIG_FOO="y"` (instead of `CONFIG_FOO=y`), or `="m"` respectively, it would still be transformed to label value `"true"`, although the label value should be `"y"`.  However, I think this is a practice this would be a non-existent corner case.

### 2. Distinguish `="y"` from `=y`

This would be the slightly hairier approach. I _think_ this would require storing a two copies of the config values in the kernel source:
1. one that contains the actual value and is used for the new `kernel.config` rules
2. one that contains the "filtered" (y|m => true) value and that would be used for the labels and legacy kconfig rule


Any thoughts @mythi ?

**Comment by marquiz** [comment]
@mythi @zvonkok @ArangoGutierrez I implemented option 2 from above (_Distinguish ="y" from =y_) in #684

**Comment by marquiz** [comment]
We need to resolve this before v0.10.0

**Comment by marquiz** [comment]
Fixed by #684 
/close
